### PR TITLE
refactor: fixes linter issues and code refinement.

### PR DESCRIPTION
**Pull request by hemani19**
**What this PR does / why we need it**:

This PR introduces linter fixes for the code base as suggested by the linter tool, [gometalinter](https://github.com/alecthomas/gometalinter). The following changes are incorporated as part of linter fixes.

- renaming of packages to avoid `UNDERSCORE` in package name.
- adding missing error checks.
- reducing cyclomatic complexity by simplifying and breaking long functions.
- adding comments to exported functions and variables.
- removing and fixing dead and ineffective code. 

The linter checking be done by executing `make check` on the codebase.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -115,7 +117,7 @@ bin/$(ARCH)/$(BIN): build-dirs
 	        ARCH=$(ARCH)                                                   \
 	        VERSION=$(VERSION)                                             \
 	        PKG=$(PKG)                                                     \
-	        ./$(PRO)/$(CMD)/build/build.sh                                               \
+	        ./$(PRO)/build/build.sh                                               \
```
Comment (kreddyj) [inline Makefile]: change this to `./$(PRO)/$(BUILD)/build.sh`

File: Makefile [hunk 2]
```diff
@@ -103,6 +104,7 @@ bin/$(ARCH)/$(BIN): build-dirs
 	    -ti                                                                \
 	    -u $$(id -u):$$(id -g)                                             \
 	    -v $$(pwd)/.go:/go:$(DOCKER_MOUNT_MODE)                            \
+		-v $$(pwd)/$(BUILD):/go/src/$(PRO)/$(BUILD):$(DOCKER_MOUNT_MODE)		\
```
Comment (kreddyj) [inline Makefile]: intendation

File: pkg/controller/controller/group_controller.go [hunk 3]
```diff
@@ -198,80 +202,96 @@ func GetAllGroups(crdclient *client.GroupCrdClient) []crd.Group {
...
+			// do nothing.
+		} else if payload.EventType == Delete {
+			existingPodDetails.EndTime = pod.GetObjectMeta().GetCreationTimestamp()
 		}
+	} else {
+		if payload.EventType == Create {
+			newPodDetails := crd.PodDetails{Name: pod.Name, StartTime: pod.GetCreationTimestamp()}
+			containers := []*crd.Container{}
+			for _, cont := range pod.Spec.Containers {
+				container := getPodMetrics(cont)
```
Comment (kreddyj) [inline pkg/controller/controller/group_controller.go]: `getContainerMetrics` or `getContainerWithMetrics` makes more sense for this function.

File: pkg/controller/controller/group_controller.go [hunk 4]
```diff
@@ -198,80 +202,96 @@ func GetAllGroups(crdclient *client.GroupCrdClient) []crd.Group {
...
+			}
+			newPodDetails.Containers = containers
+			podDetails[podKey] = &newPodDetails
+		} else if payload.EventType == Delete {
+			// do nothing.
+		}
+	}
+}
+
+func getPodMetrics(cont api_v1.Container) *crd.Container {
```
Comment (kreddyj) [inline pkg/controller/controller/group_controller.go]: `getContainerMetrics` or `getContainerWithMetrics` makes more sense for this function.

File: pkg/controller/controller/controller.go [hunk 5]
```diff
@@ -328,26 +326,16 @@ func newResourceController(client kubernetes.Interface, informer cache.SharedInd
...
 			newEvent.key, err = cache.MetaNamespaceKeyFunc(obj)
-			newEvent.eventType = "create"
+			newEvent.eventType = Create
 			newEvent.resourceType = resourceType
 			log.Printf("Processing add to %v: %s", resourceType, newEvent.key)
 			if err == nil {
 				queue.Add(newEvent)
 			}
 		},
-		UpdateFunc: func(old, new interface{}) {
```
Comment (gurusreekanthc) [inline pkg/controller/controller/controller.go]: Lets keep UpdateFunc for reference.

File: pkg/controller/controller/controller.go [hunk 6]
```diff
@@ -423,17 +411,20 @@ func (c *Controller) processItem(newEvent Event) error {
...
+	case Create:
+		str, err := json.Marshal(obj)
+		if err != nil {
+			log.Errorf("Error marshalling object %s", obj)
+		}
 		payload := &eventprocessor.Payload{Key: newEvent.key, EventType: newEvent.eventType,
 			ResourceType: newEvent.resourceType, CloudType: "aws", Data: string(str)}
 		c.conf.RingBuffer.Put(payload)
 		return nil
-	case "update":
```
Comment (gurusreekanthc) [inline pkg/controller/controller/controller.go]: Same here.

File: pkg/controller/controller/controller.go (hunk 5 above)
Comment (gurusreekanthc) [reply pkg/controller/controller/controller.go]: Perhaps we can add a TODO to have a deeper look later.

**Comment by kreddyj** [comment]
The binary of purser controller changed from `purser_controller` to just `controller`, so the docker image created will be `gurusreekanth/controller-amd64:1.0.0`. Then we need to change the image name in [custom_controller file](https://github.com/vmware/purser/blob/master/custom_controller.yaml) too.

**Comment by hemani19** [comment]
@kreddyj,  `gurusreekanth/controller-amd64:1.0.0` this name is not very intuitive. Can we come up with a better name? How is the naming done for other custom controller projects?

**Comment by kreddyj** [comment]
@hemani19, `gurusreekanth/controller-amd64:1.0.0` this is the docker image created when we do `make container`. Later we should probably get a docker account with name `purser` so that the image name will be `purser/controller:(versionNumber)`.

**Comment by kreddyj** [comment]
@gurusreekanthc can we merge this?

**Pull request merged**